### PR TITLE
feat(libflux): handle multi-file packages

### PR DESCRIPTION
**Pull request by fchikwekwe**
### Done checklist
- [ ] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: libflux/go/libflux/create_type_env.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+
+// EnvStdlib takes care of creating a flux_buffer_t, passes the buffer to
+// the Flatbuffers TypeEnvironment and then takes care of freeing the data
+func EnvStdlib() []byte {
+	var buf C.struct_flux_buffer_t
+	C.flux_get_env_stdlib(&buf)
+	defer C.flux_free(buf.data)
+	return C.GoBytes(buf.data, C.int(buf.len))
+}
```
Comment (nathanielc) [inline libflux/go/libflux/create_type_env.go]: Why was this moved to its own file?

File: libflux/go/libflux/flux.h [hunk 2]
```diff
@@ -75,6 +75,10 @@ void flux_get_env_stdlib(struct flux_buffer_t *);
 // flux_ast_pkg_t* argument, so it should not be reused after calling this function.
 struct flux_error_t *flux_analyze(struct flux_ast_pkg_t *, struct flux_semantic_pkg_t **);
 
+// flux_ast_pkg_t merges the files of a given input AST package into the file vector of a
+// given output AST package. If there are errors, then a panic will occur.
```
Comment (nathanielc) [inline libflux/go/libflux/flux.h]: We can't safely panic through the CGO boundary, it looks like the function signature is to return the error. Is this comment just out of date?

File: libflux/go/libflux/flux.h [hunk 3]
```diff
@@ -75,6 +75,10 @@ void flux_get_env_stdlib(struct flux_buffer_t *);
...
+struct flux_error_t *flux_merge_ast_pkg_files(struct flux_ast_pkg_t *, struct flux_ast_pkg_t *);
```
Comment (nathanielc) [inline libflux/go/libflux/flux.h]: We should name the args to this function since its not clear which is the input and which is the output without names.

File: libflux/go/libflux/merge.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+// #cgo CFLAGS: -I.
+// #cgo LDFLAGS: -L. -llibstd
+// #include "flux.h"
+// #include <stdlib.h>
+import "C"
+
+// Merge packages merges the files of a given input package into a given
+// output package. If there are unrecoverable errors, a panic may occur.
+func MergePackages(inPkg *ASTPkg, outPkg *ASTPkg) *ASTPkg {
```
Comment (nathanielc) [inline libflux/go/libflux/merge.go]: Should this return an error? Where does the panic come from? I don't see a call to `panic`.

File: libflux/src/flux/lib.rs [hunk 5]
```diff
@@ -223,3 +223,109 @@ pub unsafe extern "C" fn flux_error_str(err: *mut flux_error_t) -> *mut c_char {
 pub unsafe extern "C" fn flux_free(err: *mut c_void) {
     Box::from_raw(err);
 }
+
+/// flux_merge_ast_pkg_files merges the files of a given in_pkg into the file
+/// vector of a given out_pkg. If there are problems with the package identifiers,
+/// then a panic may occur.
+#[no_mangle]
+pub unsafe extern "C" fn flux_merge_ast_pkg_files(
```
Comment (nathanielc) [inline libflux/src/flux/lib.rs]: Ah, I think I understand my confusion. A panic from Rust will not result in a panic in Go. A panic in Rust will just corrupt the memory of the Go process(which may or may not panic the Go side of things). I am not sure what will happen then. We cannot panic from this Rust function. We must ensure that all errors are returned explicitly. Perhaps this is related to the bug you are looking for.

File: libflux/src/flux/lib.rs [hunk 6]
```diff
@@ -223,3 +223,109 @@ pub unsafe extern "C" fn flux_error_str(err: *mut flux_error_t) -> *mut c_char {
...
+    match merge_files(in_pkg, out_pkg) {
+        Ok(_) => std::ptr::null_mut(),
+        Err(err) => {
+            let err_handle = ErrorHandle { err: Box::new(err) };
+            Box::into_raw(Box::new(err_handle)) as *mut flux_error_t
+        }
+    }
+}
+
+pub fn merge_files(in_pkg: ast::Package, out_pkg: ast::Package) -> Result<Vec<ast::File>, Error> {
```
Comment (nathanielc) [inline libflux/src/flux/lib.rs]: nit, can you rename this to `merge_packages`?

File: libflux/go/libflux/flux.h (hunk 2 above)
Comment (fchikwekwe) [reply libflux/go/libflux/flux.h]: Yeah this comment is out of date.

File: libflux/go/libflux/merge.go (hunk 4 above)
Comment (fchikwekwe) [reply libflux/go/libflux/merge.go]: Yeah, another comment out of date.

File: libflux/src/flux/lib.rs (hunk 5 above)
Comment (fchikwekwe) [reply libflux/src/flux/lib.rs]: Yeah, there aren't any panics anymore. I'm just propagating errors now. I was panicking originally until I started trying to glue the Go and Rust sides together.

File: libflux/go/libflux/analyze_test.go [hunk 7]
```diff
@@ -56,3 +58,170 @@ func TestAnalyze(t *testing.T) {
...
+
+	err := libflux.MergePackages(pkg1, pkg2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	gotFB, err := pkg2.MarshalFB()
+	got := ast.DeserializeFromFlatBuffer(gotFB)
+
+	if err != nil {
```
Comment (nathanielc) [inline libflux/go/libflux/analyze_test.go]: nit, this error check should come before the call to DesierlizFromFlatBuffer.

File: libflux/go/libflux/analyze.go [hunk 8]
```diff
@@ -76,13 +76,31 @@ func Analyze(astPkg *ASTPkg) (*SemanticPkg, error) {
 	return p, nil
 }
 
+// Merge packages merges the files of a given input package into a given output package.
+// This function borrows the input and output packages, but does not own them. Memory
+// must still be freed by the caller of this function.
+func MergePackages(inPkg *ASTPkg, outPkg *ASTPkg) error {
```
Comment (wolffcm) [inline libflux/go/libflux/analyze.go]: I think this function would better located in `parser.go`, since it's dealing with ASTs rather than semantic graphs.

File: libflux/go/libflux/analyze_test.go [hunk 9]
```diff
@@ -2,6 +2,8 @@ package libflux_test
 
 import (
 	"errors"
+	"github.com/influxdata/flux/ast"
```
Comment (wolffcm) [inline libflux/go/libflux/analyze_test.go]: *nit* these imports are out of order.

File: libflux/go/libflux/analyze_test.go (hunk 7 above)
Comment (fchikwekwe) [reply libflux/go/libflux/analyze_test.go]: true, good catch.

File: libflux/go/libflux/analyze_test.go [hunk 10]
```diff
@@ -56,3 +58,170 @@ func TestAnalyze(t *testing.T) {
...
+		Files: []*ast.File{
+			{
+				Name:     "",
+				Metadata: "parser-type=rust",
+				Package: &ast.PackageClause{
+					Name: &ast.Identifier{Name: "foo"},
+				},
+				Body: []ast.Statement{
+					&ast.VariableAssignment{
+						ID:   &ast.Identifier{Name: "b"},
```
Comment (wolffcm) [inline libflux/go/libflux/analyze_test.go]: Hmm, given the way the inputs are arranged, I would have expected the file with `a = 1` to appear first, and `b = 1` to be second.  I wonder if `MergePackages` should just append all the files in the second argument to the files in the first argument?  This way the order of the files would be preserved.

File: libflux/go/libflux/analyze_test.go [hunk 11]
```diff
@@ -56,3 +58,170 @@ func TestAnalyze(t *testing.T) {
...
+	want := &ast.Package{
+		Package: "foo",
+		Files: []*ast.File{
+			{
+				Name:     "",
+				Metadata: "parser-type=rust",
+				Package: &ast.PackageClause{
+					Name: &ast.Identifier{Name: "foo"},
+				},
+				Body: []ast.Statement{
```
Comment (wolffcm) [inline libflux/go/libflux/analyze_test.go]: I'm confused by how this test function is named, since there does seem to be some input files here.  I wonder if you parsed the empty string `""`, is that what you're trying to test?

File: libflux/go/libflux/flux.h [hunk 12]
```diff
@@ -75,6 +75,11 @@ void flux_get_env_stdlib(struct flux_buffer_t *);
 // flux_ast_pkg_t* argument, so it should not be reused after calling this function.
 struct flux_error_t *flux_analyze(struct flux_ast_pkg_t *, struct flux_semantic_pkg_t **);
 
+// flux_merge_ast_pkgs merges the files of a given input AST package into the file vector of a
+// given output AST package. This function borrows the packages, but it does not own them. The
+// caller of this function still needs to free the package memory on the Go side.
+struct flux_error_t *flux_merge_ast_pkgs(struct flux_ast_pkg_t *, struct flux_ast_pkg_t *);
```
Comment (wolffcm) [inline libflux/go/libflux/flux.h]: *nit* I would move this prototype closer to the other functions that deal with `flux_ast_pkg_t*` (like parsing) just to keep related functions near to each other.

File: libflux/go/libflux/analyze_test.go [hunk 13]
```diff
@@ -56,3 +58,170 @@ func TestAnalyze(t *testing.T) {
...
+	if !cmp.Equal(got, want, asttest.IgnoreBaseNodeOptions...) {
+		t.Errorf(
+			"MergePackages unexpected packages -want/+got:\n %v",
+			cmp.Diff(want, got, asttest.IgnoreBaseNodeOptions...),
+		)
+	}
+}
+
+func TestNoOutFiles(t *testing.T) {
+	pkg1 := libflux.Parse(`
```
Comment (wolffcm) [inline libflux/go/libflux/analyze_test.go]: There's some boilerplate here that's repeated for each test method, you could use an array of test cases and loop over them to be more concise.

File: libflux/go/libflux/analyze_test.go [hunk 14]
```diff
@@ -56,3 +58,170 @@ func TestAnalyze(t *testing.T) {
...
+		},
+	}
+
+	if !cmp.Equal(got, want, asttest.IgnoreBaseNodeOptions...) {
+		t.Errorf(
+			"MergePackages unexpected packages -want/+got:\n %v",
+			cmp.Diff(want, got, asttest.IgnoreBaseNodeOptions...),
+		)
+	}
+}
```
Comment (wolffcm) [inline libflux/go/libflux/analyze_test.go]: It would be good to have some tests for the errors that can happen, like when package names do not match, etc.

File: libflux/go/libflux/analyze_test.go (hunk 11 above)
Comment (fchikwekwe) [reply libflux/go/libflux/analyze_test.go]: Well parsing the empty string actually throws an error on the rust side because that package would have no package clause. 

Perhaps these "empty packages" tests are obsolete now that I've eliminated the logic that I wrote for the `ParseDirToHandle` function in my last iteration.

Comment (fchikwekwe) [reply libflux/go/libflux/analyze_test.go]: Yeah, I think I'm just gonna replace those two tests. They don't really make sense anymore.

File: libflux/go/libflux/analyze_test.go (hunk 13 above)
Comment (fchikwekwe) [reply libflux/go/libflux/analyze_test.go]: Yes,  I did try that but it was causing an issue in the test information. I wasn't sure if conciseness was worth spending time on another bug.

File: libflux/go/libflux/analyze_test.go (hunk 14 above)
Comment (fchikwekwe) [reply libflux/go/libflux/analyze_test.go]: I have some test like that on the Rust side. I can add some similar tests here if that would be helpful.

Comment (wolffcm) [reply libflux/go/libflux/analyze_test.go]: I think it would be good to test the error path.  There is some code on the Go side that will not get exercised otherwise.

File: libflux/go/libflux/analyze_test.go (hunk 13 above)
Comment (fchikwekwe) [reply libflux/go/libflux/analyze_test.go]: Nevermind. Fixed this.

File: libflux/go/libflux/analyze_test.go (hunk 14 above)
Comment (fchikwekwe) [reply libflux/go/libflux/analyze_test.go]: I made some changes to the test file. Let me know if you know of any other cases that I should also cover.
